### PR TITLE
Accelerometer Support #569

### PR DESCRIPTION
**Pull request by kevvz**
#569 

How to Test:

1. Flash the zephyr.uf2 file located in friend/firmware/firmwarev1.0/accel_build.
2. Run the updated app in debug mode and open the console.
3. Wait for the hardware to connect. Then after a while you should see accelerometer data, including the gyroscope appear on the console. 
4.  Throw your friend hard, preferably onto a soft surface like a couch. If you throw hard enough you should get a notification labeled 'ouch.....' . I've tried fake falling and if the parameter fall_number listed in communication.dart is low enough, you can trigger the notification but you may trigger it on accident if you move fast.

### CODE REVIEW

**Comment by josancamon19** [comment]
Awesome, while we wait for gyroscope data, can you do in mobile smth similar that detects changes for it, and prints something?

And together with gyroscope the person fell example?

**Comment by josancamon19** [comment]
In the meantime, pls conver this to a Draft PR @kevvz

**Comment by kevvz** [comment]
Added. It will printing in the debug console if you download the uf2 file provided in the PR.

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 1]
```diff
@@ -65,7 +68,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+static struct bt_gatt_attr accel_service_attr[] = {
+    BT_GATT_PRIMARY_SERVICE(&accel_uuid),//primary description
+    BT_GATT_CHARACTERISTIC(&accel_uuid_x.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ, accel_data_read_characteristic, NULL, NULL),//data type
+    BT_GATT_CCC(accel_ccc_config_changed_handler, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),//scheduler
+};
+static struct bt_gatt_service accel_service = BT_GATT_SERVICE(accel_service_attr);
+
+static ssize_t accel_data_read_characteristic(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf, uint16_t len, uint16_t offset)
+{
+    printk("audio_data_read_characteristic\n");
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Erhm... :-)

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 2]
```diff
@@ -65,7 +68,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+    int axis_mode = 6; //3 for accel, 6 for (also) gyro
+    return bt_gatt_attr_read(conn, attr, buf, len, offset, &axis_mode, sizeof(axis_mode));
+}
+
+
+#define ACCEL_REFRESH_INTERVAL 500 // 0.5 seconds
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Can we bump this to 100ms? (as long as it doesn't interfere with the audio and video data streams)

**Comment by francip** [comment]
Btw, you'll have to merge with https://github.com/BasedHardware/Omi/pull/590

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 3]
```diff
@@ -65,7 +68,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+	struct sensor_value a_z;
+    struct sensor_value g_x;
+    struct sensor_value g_y;
+    struct sensor_value g_z;
+
+};
+static struct sensors mega_sensor;
+static struct device *lsm6dsl_dev;
+//Arbritrary uuid, feel free to change
+static struct bt_uuid_128 accel_uuid = BT_UUID_INIT_128(BT_UUID_128_ENCODE(0x00000000,0x0000,0x1000,0x7450,0xBE2E44B06B00));
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: How did you generate this UUID? Looks... suspicious to me :-)
Also, any particular reason we want this in a separate service?

Comment (kevvz) [reply Friend/firmware/firmware_v1.0/src/transport.c]: I was following the conventions that appeared to be in the file already. The contributors separated the services such as battery, firmware ota, and audio, so I thought it would be appropriate to make a new service too. I suppose you could stuff all data into a single service labeled "friends", although I'm unsure if it is possible.

File: Friend/firmware/firmware_v1.0/src/transport.c (hunk 1 above)
Comment (kevvz) [reply Friend/firmware/firmware_v1.0/src/transport.c]: Just in case I wanted to add more features!

File: Friend/firmware/firmware_v1.0/src/transport.c (hunk 2 above)
Comment (kevvz) [reply Friend/firmware/firmware_v1.0/src/transport.c]: Sure. I believe the main intention was for pendant gestures, so if we need to go back to it we can.

**Comment by kevvz** [comment]
Merging the firmware appears to have given me a whole slew of problems. I can't seem to even connect to the BLE scanner app anymore. Give me a little more time.

**Comment by kevvz** [comment]
I redownloaded the repository from Basedhardware/Ori, and the built firmware causes the device to crash on the bluetooth connection. I believe there is something wrong in the firmware and merging doesn't appear to solve anything. Would it be a good idea to rollback until the firmware is fixed? I think #590 may be bugged unless I get someone else's firmware

**Comment by kevvz** [comment]
Alright, I kinda found the problem. If I replace the Prj.conf file in main with the one I used to test my accelerometer, it appears to work just fine. There is something in the file that breaks the device, although there are so many that it is hard to pinpoint the exact one.

File: Friend/firmware/firmware_v1.0/src/transport.c (hunk 3 above)
Comment (francip) [reply Friend/firmware/firmware_v1.0/src/transport.c]: We have separate services, because Battery Service and Device Information Service are well-known services that are defined on the BLE standard with 16-bit service IDs.

The OTA is a custom service as well, but it is a well-known Nordic service, so it is a de-facto industry standard.

The audio and video data streaming are custom logic specific to our device. So are the accelerometer and button. We could be putting these into separate services, but honestly the only reason would be to make it easy to build devices that have accelerometer/button, but don't have audio/video streaming, which I don't see happening.

Separately, there's a limitation on the number of custom services that can be advertised in the broadcast advertising packets, which makes the discovery of multiple services bit more complex than a single service. This has to do with the fact that the space in the advertising packets is limited, and custom services have 128-bit UUIDs.

And from app developers POV, it might be easier to maintain state transition if dealing with a single service with multiple characteristics, than with multiple services.

Honestly, neither of these is a strong argument against separate service on its own, but all together make me lean towards merging accelerometer/button into the main service together with the audio/video data.

But I am not going to fight this too hard, as long as the actual logic works :-)

**Comment by francip** [comment]
@koconder can you look at what might be happening here? :-/

**Comment by vincentkoc** [comment]
@kevvz your changes look OK, do you need me to check or did your last changes fix this? - It was likely the speaker/audio one, which is all good. And there was a line disabling certain components.

**Comment by kevvz** [comment]
@koconder Everything works fine for me. I only needed to change the overlay file name and replaced the prj.conf file with my version.

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf [hunk 4]
```diff
@@ -34,7 +34,7 @@ CONFIG_BT_PHY_UPDATE=y
 CONFIG_BT_BAS=y
 CONFIG_BT_DIS=y
 CONFIG_BT_DIS_PNP=n
-CONFIG_BT_DIS_MODEL="Friend DevKit 2"
+CONFIG_BT_DIS_MODEL="Friend"
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf]: Revert this change, please.

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf [hunk 5]
```diff
@@ -55,25 +55,23 @@ CONFIG_BT_AUTO_DATA_LEN_UPDATE=y
...
-CONFIG_LOG_PRINTK=y
-CONFIG_LOG_MODE_IMMEDIATE=y
-CONFIG_SERIAL=y
-CONFIG_UART_CONSOLE=y
-CONFIG_LOG_BACKEND_UART=y
-CONFIG_LOG_BACKEND_UART_OUTPUT_TEXT=y
-CONFIG_LOG_DEFAULT_LEVEL=3
-
-# Debug (This value breaks some builds)
+# CONFIG_DEBUG=y
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf]: Revert this as well.

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf [hunk 6]
```diff
@@ -95,58 +93,4 @@ CONFIG_LOG_DEFAULT_LEVEL=3
 # CONFIG_SEGGER_RTT_BUFFER_SIZE_UP=2048
 # CONFIG_RTT_CONSOLE=y
 # CONFIG_LOG_BACKEND_RTT=y
-CONFIG_CODEC_OPUS=y
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf]: And this.

File: Friend/firmware/firmware_v1.0/src/transport.c (hunk 1 above)
Comment (francip) [reply Friend/firmware/firmware_v1.0/src/transport.c]: Lol, I should've been more clear. Copy/paste. I was trying to hint that the printk should actually say "accel_data_read_characteristic" :-)

On a more serious note, change all printk to LOG to make it consistent with the rest of the code that was changed in #590 .

File: app/lib/pages/home/device.dart [hunk 7]
```diff
@@ -14,6 +14,8 @@ class ConnectedDevice extends StatefulWidget {
   final BTDeviceStruct? device;
   final int batteryLevel;
 
+
```
Comment (francip) [inline app/lib/pages/home/device.dart]: nitpick: revert

**Comment by francip** [comment]
Saw few changes that slipped through the merge, which don't seem to be related to your code. Please revert them.
Also, few other nitpicks here and there.
Also, can you generate a different UUID for the new service? This one looks too suspicious to me. :-)

I can't comment on the Flutter code much. It looked reasonable to me, but that language is not my native, so...

Anyway, I think we're pretty close to getting this in.

**Comment by francip** [comment]
And sorry I've been too nitpicky

**Review by kevvz — COMMENTED** [review]
Finished requests

**Comment by kevvz** [comment]
> And sorry I've been too nitpicky

No worries, thanks for taking the time to read my code!

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 8]
```diff
@@ -1,4 +1,3 @@
-#include <zephyr/logging/log.h>
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Why are you removing this?

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv1.conf [hunk 9]
```diff
@@ -55,12 +55,6 @@ CONFIG_BT_AUTO_DATA_LEN_UPDATE=y
 CONFIG_BT_USER_PHY_UPDATE=y
 CONFIG_BT_AUTO_PHY_UPDATE=y
 
-#accelerometer
-CONFIG_LSM6DSL=y
-CONFIG_LSM6DSL_ENABLE_TEMP=n
-CONFIG_LSM6DSL_TRIGGER_GLOBAL_THREAD=y
-CONFIG_SENSOR=y
```
Comment (francip) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv1.conf]: I think this (and similar in other .conf files) shouldn't have been reverted

**Comment by francip** [comment]
I think you reverted accidentally more than you should've :-(

**Review by kevvz — COMMENTED** [review]
revised some files

**Comment by vincentkoc** [comment]
Once this is resolved and merged ill update the overlay file for Audio and SD Card, just waiting to avoid merge conflicts.

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf [hunk 10]
```diff
@@ -16,7 +16,7 @@ CONFIG_ADC=y
 
 CONFIG_BT=y
 CONFIG_BT_PERIPHERAL=y
-CONFIG_BT_DEVICE_NAME="Friend"
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf:19]: Revert this change

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv1.conf [hunk 11]
```diff
@@ -90,6 +90,11 @@ CONFIG_LOG_DEFAULT_LEVEL=3
 # CONFIG_BT_RX_STACK_SIZE=4096
 # CONFIG_BT_CTLR_RX_PRIO_STACK_SIZE=4096
 
+CONFIG_LSM6DSL=y
+CONFIG_LSM6DSL_ENABLE_TEMP=n
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv1.conf:94]: How comes some of these are not on the devkit2 proj file? If not needed remove or add to both files

```
CONFIG_LSM6DSL_ENABLE_TEMP=n
CONFIG_LSM6DSL_TRIGGER_GLOBAL_THREAD=y
CONFIG_SENSOR=y
```

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 12]
```diff
@@ -67,7 +70,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+
+    sensor_sample_fetch_chan(lsm6dsl_dev, SENSOR_CHAN_GYRO_XYZ);
+	sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_X, &mega_sensor.g_x);
+	sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_Y, &mega_sensor.g_y);
+	sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_Z, &mega_sensor.g_z);
+
+   //only time mega sensor is changed is through here (hopefully),  so no chance of race condition
+    int err = bt_gatt_notify(current_connection, &accel_service.attrs[1], &mega_sensor, sizeof(mega_sensor));
+    if (err) {
+       LOG_INF("Error updating accelerometer data\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 13]
```diff
@@ -67,7 +70,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+        LOG_INF("Invalid CCC value: %u", value);
+    }
+}
+
+int accel_start() {
+    struct sensor_value odr_attr;
+    lsm6dsl_dev = DEVICE_DT_GET_ONE(st_lsm6dsl);
+    k_msleep(50);
+    if (lsm6dsl_dev == NULL) {
+        LOG_INF("Could not get LSM6DSL device\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR` also remove all the `\n` from all the logs its not needed and polutes the log

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 14]
```diff
@@ -67,7 +70,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+        return 0;
+	}
+    if (!device_is_ready(lsm6dsl_dev)) {
+		LOG_INF("LSM6DSL: not ready\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 15]
```diff
@@ -67,7 +70,119 @@ static struct bt_gatt_attr dfu_service_attr[] = {
...
+		    SENSOR_ATTR_SAMPLING_FREQUENCY, &odr_attr) < 0) {
+	LOG_INF("Cannot set sampling frequency for gyro.\n");
+	    return 0;
+	}
+    if (sensor_sample_fetch(lsm6dsl_dev) < 0) {
+    LOG_INF("Sensor sample update error\n");
+    return 0;
+	}
+
+    LOG_INF("accelerometer is ready for use \n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Case, `Accelerometer`

File: Friend/firmware/firmware_v1.0/src/transport.c [hunk 16]
```diff
@@ -453,6 +567,15 @@ int transport_start()
     }
     LOG_INF("Transport bluetooth initialized");
 
+    err = accel_start();
+
+    if (!err) {
+        LOG_INF("accelerometer failed to activate\n");
```
Comment (vincentkoc) [inline Friend/firmware/firmware_v1.0/src/transport.c]: Use `LOG_ERR` and also ensure correct case `Accelerometer`

**Review by vincentkoc — CHANGES_REQUESTED** [review]
Please also nuke the .uf2 file from the merge

File: Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf (hunk 10 above)
Comment (kevvz) [reply Friend/firmware/firmware_v1.0/prj_xiao_ble_sense_devkitv2-adafruit.conf:19]: Francis told me to have it named Friend. Im getting mixed signals here?

**Review by kevvz — COMMENTED** [review]
done

**Comment by kodjima33** [comment]
@kevvz 
CleanShot 2024-08-24 at 15 16 45

cc @josancamon19

**Review by josancamon19 — APPROVED** [review]
